# Hazem Zaky
# Programming Exercise \#1
---

# Preamble

In [85]:
# optional code cell when using Google Colab with Google Drive

# remove the docstring comment block below in order to mount Google Drive
'''
# mount Google Drive in Google Colab
from google.colab import drive
drive.mount('/content/drive')

# change directory using the magic command %cd
### replace [MY PATH] below with your own path in Google Drive ###
### %cd /content/drive/My\ Drive/[MY PATH] ###
'''

"\n# mount Google Drive in Google Colab\nfrom google.colab import drive\ndrive.mount('/content/drive')\n\n# change directory using the magic command %cd\n### replace [MY PATH] below with your own path in Google Drive ###\n### %cd /content/drive/My\\ Drive/[MY PATH] ###\n"

In [86]:
# import relevant Python libraries

import numpy as np
import pandas as pd
from scipy import stats as sps
from IPython.display import display, Latex

# **1. Fetal Health Classification Dataset**

## **Clean Dataset**

### Problem 1.1

In [87]:
# load the clean dataset csv file into a pandas dataframe

fetal_df = pd.read_csv('data/fetal_health_dataset_clean.csv') 

#### (a)

This is a supervised machine learning task because we are given the output of each data value as whether the fetus is normal, suspect, or pathological.

#### (b)

In [88]:
print('Axes: ', fetal_df.axes)
print('\n\n')
print('Dtypes: ', fetal_df.dtypes)

Axes:  [RangeIndex(start=0, stop=2126, step=1), Index(['baseline value', 'accelerations', 'fetal_movement',
       'uterine_contractions', 'light_decelerations', 'severe_decelerations',
       'prolongued_decelerations', 'abnormal_short_term_variability',
       'mean_value_of_short_term_variability',
       'percentage_of_time_with_abnormal_long_term_variability',
       'mean_value_of_long_term_variability', 'histogram_width',
       'histogram_min', 'histogram_max', 'histogram_number_of_peaks',
       'histogram_number_of_zeroes', 'histogram_mode', 'histogram_mean',
       'histogram_median', 'histogram_variance', 'histogram_tendency',
       'fetal_health'],
      dtype='object')]



Dtypes:  baseline value                                              int64
accelerations                                             float64
fetal_movement                                            float64
uterine_contractions                                      float64
light_decelerations           

#### (c)

In [89]:
fetal_df.head(10)

baseline value  accelerations  fetal_movement  uterine_contractions  \
0             120          0.000             0.0                 0.000   
1             132          0.006             0.0                 0.006   
2             133          0.003             0.0                 0.008   
3             134          0.003             0.0                 0.008   
4             132          0.007             0.0                 0.008   
5             134          0.001             0.0                 0.010   
6             134          0.001             0.0                 0.013   
7             122          0.000             0.0                 0.000   
8             122          0.000             0.0                 0.002   
9             122          0.000             0.0                 0.003   

   light_decelerations  severe_decelerations  prolongued_decelerations  \
0                0.000                   0.0                     0.000   
1                0.003                   0.0                     0.000   
2                0.003                   0.0                     0.000   
3                0.003                   0.0                     0.000   
4                0.000                   0.0                     0.000   
5                0.009                   0.0                     0.002   
6                0.008                   0.0                     0.003   
7                0.000                   0.0                     0.000   
8                0.000                   0.0                     0.000   
9                0.000                   0.0                     0.000   

   abnormal_short_term_variability  mean_value_of_short_term_variability  \
0                               73                                   0.5   
1                               17                                   2.1   
2                               16                                   2.1   
3                               16                                   2.4   
4                               16                                   2.4   
5                               26                                   5.9   
6                               29                                   6.3   
7                               83                                   0.5   
8                               84                                   0.5   
9                               86                                   0.3   

   percentage_of_time_with_abnormal_long_term_variability  ...  histogram_min  \
0                                                 43       ...             62   
1                                                  0       ...             68   
2                                                  0       ...             68   
3                                                  0       ...             53   
4                                                  0       ...             53   
5                                                  0       ...             50   
6                                                  0       ...             50   
7                                                  6       ...             62   
8                                                  5       ...             62   
9                                                  6       ...             62   

   histogram_max  histogram_number_of_peaks  histogram_number_of_zeroes  \
0            126                          2                           0   
1            198                          6                           1   
2            198                          5                           1   
3            170                         11                           0   
4            170                          9                           0   
5            200                          5                           3   
6            200                          6                           3   
7            130                          0 

#### (d)

In [90]:
fetal_df.shape

(2126, 22)

#### (e)

_[Your answer for 1.1(e) goes here]_

#### (f)

There are 2126 total samples.

#### (g)

21 independent variables.
baseline value
accelerations
fetal_movement
uterine_contractions
light_decelerations
severe_decelerations
prolongued_decelerations
abnormal_short_term_variability
mean_value_of_short_term_variability
percentage_of_time_with_abnormal_long_term_variability
mean_value_of_long_term_variability
histogram_width
histogram_min
histogram_max
histogram_number_of_peaks
histogram_number_of_zeroes
histogram_mode
histogram_mean
histogram_median
histogram_variance
histogram_tendency


#### (h)

1 Dependent variable
fetal_health

#### (i)

n = 2126, p = 21


#### (j)

n = 2126, m = 1

#### (k)

Some of the variables in the dataset are preprocessed variables, and some are raw variables. This is because some of the variables, such as histogram_tendency, histogram_variance, etc. Are based on the general data obtained from the histograms the raw data produced, whereas some variables such as accelerations, fetal_movement, etc. are variables that are obtained directly from the raw data.

#### (l)

In [91]:
for column in fetal_df.columns:
    unique_values = pd.unique(fetal_df[column])
    print(f"Unique values in '{column}': {unique_values}\n")

Unique values in 'baseline value': [120 132 133 134 122 151 150 131 130 129 128 124 115 114 116 158 156 148
 149 146 144 142 136 141 138 140 154 145 139 125 123 159 143 119 121 127
 126 118 135 137 147 157 117 152 112 106 110 160]

Unique values in 'accelerations': [0.    0.006 0.003 0.007 0.001 0.005 0.009 0.002 0.008 0.004 0.01  0.015
 0.013 0.014 0.011 0.017 0.012 0.016 0.019 0.018]

Unique values in 'fetal_movement': [0.    0.072 0.222 0.408 0.38  0.441 0.383 0.451 0.469 0.34  0.425 0.334
 0.135 0.099 0.108 0.112 0.089 0.103 0.085 0.109 0.079 0.065 0.055 0.058
 0.047 0.038 0.012 0.018 0.02  0.005 0.003 0.006 0.001 0.004 0.009 0.01
 0.002 0.008 0.007 0.028 0.026 0.107 0.013 0.016 0.029 0.05  0.053 0.011
 0.015 0.022 0.021 0.017 0.019 0.025 0.014 0.024 0.023 0.035 0.054 0.03
 0.048 0.088 0.043 0.052 0.091 0.033 0.092 0.084 0.115 0.041 0.04  0.027
 0.031 0.063 0.06  0.071 0.306 0.298 0.139 0.189 0.157 0.235 0.36  0.455
 0.443 0.47  0.477 0.446 0.481 0.369 0.335 0.43  0.346 0.323 0.375

histogram_tendency
fetal_health

These are the only categorical variables because the rest of the variables have continuous values that do not have specific bins.


#### (m)

The histogram_tendency variable goes by nominal encoding, where the numbers don't have a specific meaning. The fetal_health variable goes by ordinal encoding because the unique values represent categories of fetal health, with 1 being the healthiest and 3 being the least healthy.

#### (n)

In [92]:
normal = fetal_df[fetal_df['fetal_health'] == 1].count().values[0]
suspect = fetal_df[fetal_df['fetal_health'] == 2].count().values[0]
pathological = fetal_df[fetal_df['fetal_health'] == 3].count().values[0]
print("Normal: ",normal)
print("Suspect: ",suspect)
print("Pathological: ",pathological)


Normal:  1655
Suspect:  295
Pathological:  176


## **Dirty Dataset**

In [93]:
# load the dirty dataset csv file into a pandas dataframe

fetal_dirty_df = pd.read_csv('data/fetal_health_dataset_dirty.csv') 

### Problem 1.2

#### (a)

In [94]:
fetal_dirty_df.isna().sum()

baseline value                                            47
accelerations                                              0
fetal_movement                                             0
uterine_contractions                                       0
light_decelerations                                        0
severe_decelerations                                       0
prolongued_decelerations                                   0
abnormal_short_term_variability                            0
mean_value_of_short_term_variability                       0
percentage_of_time_with_abnormal_long_term_variability    10
mean_value_of_long_term_variability                        0
histogram_width                                            0
histogram_min                                              0
histogram_max                                             24
histogram_number_of_peaks                                  0
histogram_number_of_zeroes                                 0
histogram_mode          

#### (b)

In [95]:
fetal_dirty_df.columns[fetal_dirty_df.isna().sum() != 0].tolist()

['baseline value',
 'percentage_of_time_with_abnormal_long_term_variability',
 'histogram_max',
 'histogram_mode']

#### (c)

In [96]:
fetal_dirty_df[fetal_dirty_df.isna().any(axis=1)].shape[0]

86

### Problem 1.3

In [114]:
fetal_dirty_df.loc[fetal_dirty_df['baseline value'] < 106,'baseline value'] = np.nan
fetal_dirty_df.loc[fetal_dirty_df['baseline value'] > 160,'baseline value'] = np.nan

fetal_dirty_df.loc[fetal_dirty_df['accelerations'] < 0,'accelerations'] = np.nan
fetal_dirty_df.loc[fetal_dirty_df['accelerations'] > 0.02,'accelerations'] = np.nan

fetal_dirty_df.loc[fetal_dirty_df['fetal_movement'] < 0,'fetal_movement'] = np.nan
fetal_dirty_df.loc[fetal_dirty_df['fetal_movement'] > 0.48,'fetal_movement'] = np.nan

fetal_dirty_df.loc[fetal_dirty_df['uterine_contractions'] < 0,'uterine_contractions'] = np.nan
fetal_dirty_df.loc[fetal_dirty_df['uterine_contractions'] > 0.01,'uterine_contractions'] = np.nan

fetal_dirty_df.loc[fetal_dirty_df['light_decelerations'] < 0,'light_decelerations'] = np.nan
fetal_dirty_df.loc[fetal_dirty_df['light_decelerations'] > 0.01,'light_decelerations'] = np.nan

fetal_dirty_df.loc[fetal_dirty_df['severe_decelerations'] < 0,'severe_decelerations'] = np.nan
fetal_dirty_df.loc[fetal_dirty_df['severe_decelerations'] > 0,'severe_decelerations'] = np.nan

fetal_dirty_df.loc[fetal_dirty_df['prolongued_decelerations'] < 0,'prolongued_decelerations'] = np.nan
fetal_dirty_df.loc[fetal_dirty_df['prolongued_decelerations'] > 0.01,'prolongued_decelerations'] = np.nan

fetal_dirty_df.loc[fetal_dirty_df['abnormal_short_term_variability'] < 12,'abnormal_short_term_variability'] = np.nan
fetal_dirty_df.loc[fetal_dirty_df['abnormal_short_term_variability'] > 87,'abnormal_short_term_variability'] = np.nan

fetal_dirty_df.loc[fetal_dirty_df['mean_value_of_short_term_variability'] < 0.2,'mean_value_of_short_term_variability'] = np.nan
fetal_dirty_df.loc[fetal_dirty_df['mean_value_of_short_term_variability'] > 7,'mean_value_of_short_term_variability'] = np.nan

fetal_dirty_df.loc[fetal_dirty_df['percentage_of_time_with_abnormal_long_term_variability'] < 0,'percentage_of_time_with_abnormal_long_term_variability'] = np.nan
fetal_dirty_df.loc[fetal_dirty_df['percentage_of_time_with_abnormal_long_term_variability'] > 91,'percentage_of_time_with_abnormal_long_term_variability'] = np.nan

fetal_dirty_df.loc[fetal_dirty_df['mean_value_of_long_term_variability'] < 0,'mean_value_of_long_term_variability'] = np.nan
fetal_dirty_df.loc[fetal_dirty_df['mean_value_of_long_term_variability'] > 50.7,'mean_value_of_long_term_variability'] = np.nan

fetal_dirty_df.loc[fetal_dirty_df['histogram_width'] < 3,'histogram_width'] = np.nan
fetal_dirty_df.loc[fetal_dirty_df['histogram_width'] > 180,'histogram_width'] = np.nan

fetal_dirty_df.loc[fetal_dirty_df['histogram_min'] < 50,'histogram_min'] = np.nan
fetal_dirty_df.loc[fetal_dirty_df['histogram_min'] > 159,'histogram_min'] = np.nan

fetal_dirty_df.loc[fetal_dirty_df['histogram_max'] < 122,'histogram_max'] = np.nan
fetal_dirty_df.loc[fetal_dirty_df['histogram_max'] > 238,'histogram_max'] = np.nan

fetal_dirty_df.loc[fetal_dirty_df['histogram_number_of_peaks'] < 0,'histogram_number_of_peaks'] = np.nan
fetal_dirty_df.loc[fetal_dirty_df['histogram_number_of_peaks'] > 18,'histogram_number_of_peaks'] = np.nan

fetal_dirty_df.loc[fetal_dirty_df['histogram_number_of_zeroes'] < 0,'histogram_number_of_zeroes'] = np.nan
fetal_dirty_df.loc[fetal_dirty_df['histogram_number_of_zeroes'] > 10,'histogram_number_of_zeroes'] = np.nan

fetal_dirty_df.loc[fetal_dirty_df['histogram_mode'] < 60,'histogram_mode'] = np.nan
fetal_dirty_df.loc[fetal_dirty_df['histogram_mode'] > 187,'histogram_mode'] = np.nan

fetal_dirty_df.loc[fetal_dirty_df['histogram_mean'] < 73,'histogram_mean'] = np.nan
fetal_dirty_df.loc[fetal_dirty_df['histogram_mean'] > 182,'histogram_mean'] = np.nan

fetal_dirty_df.loc[fetal_dirty_df['histogram_median'] < 77,'histogram_median'] = np.nan
fetal_dirty_df.loc[fetal_dirty_df['histogram_median'] > 186,'histogram_median'] = np.nan

fetal_dirty_df.loc[fetal_dirty_df['histogram_variance'] < 0,'histogram_variance'] = np.nan
fetal_dirty_df.loc[fetal_dirty_df['histogram_variance'] > 269,'histogram_variance'] = np.nan

fetal_dirty_df.loc[fetal_dirty_df['histogram_tendency'] < -1,'histogram_tendency'] = np.nan
fetal_dirty_df.loc[fetal_dirty_df['histogram_tendency'] > 1,'histogram_tendency'] = np.nan

fetal_dirty_df.loc[fetal_dirty_df['fetal_health'] < 1,'fetal_health'] = np.nan
fetal_dirty_df.loc[fetal_dirty_df['fetal_health'] > 3,'fetal_health'] = np.nan
fetal_dirty_df.loc[91:100]

baseline value  accelerations  fetal_movement  uterine_contractions  \
91            148.0          0.000           0.000                 0.000   
92            148.0          0.000           0.003                 0.000   
93              NaN          0.000           0.000                 0.000   
94            148.0          0.000           0.005                 0.002   
95            148.0          0.000           0.000                 0.000   
96            148.0          0.000           0.000                 0.000   
97            148.0          0.000           0.001                 0.000   
98            148.0          0.000           0.001                 0.000   
99            125.0          0.010           0.007                 0.002   
100           125.0          0.005           0.000                 0.002   

     light_decelerations  severe_decelerations  prolongued_decelerations  \
91                 0.000                   0.0                       0.0   
92                 0.000                   0.0                       0.0   
93                 0.000                   0.0                       0.0   
94                 0.000                   0.0                       0.0   
95                 0.000                   0.0                       0.0   
96                 0.000                   0.0                       0.0   
97                 0.001                   0.0                       0.0   
98                 0.001                   0.0                       0.0   
99                 0.000                   0.0                       0.0   
100                0.003                   0.0                       0.0   

     abnormal_short_term_variability  mean_value_of_short_term_variability  \
91                              67.0                                   0.4   
92                              70.0                                   0.3   
93                              68.0                                   0.3   
94                              72.0                                   0.3   
95                              58.0                                   0.5   
96                              63.0                                   0.4   
97                              60.0                                   0.5   
98                              61.0                                   0.5   
99                              19.0                                   1.9   
100                             25.0                                   1.7   

     percentage_of_time_with_abnormal_long_term_variability  ...  \
91                                                67.0       ...   
92                                                68.0       ...   
93                                                75.0       ...   
94                                                74.0       ...   
95                                                 7.0       ...   
96                                                30.0       ...   
97                                                49.0       ...   
98                                                39.0       ...   
99                                                 0.0       ...   
100                                                6.0       ...   

     histogram_min  histogram_max  histogram_number_of_peaks  \
91           139.0          154.0                        4.0   
92           136.0          153.0                        2.0   
93           128.0            NaN                        3.0   
94           136.0          152.0                        0.0   
95           135.0          156.0                        1.0   
96           121.0          156.0                        3.0   
97            78.0          161.0                        3.0   
98           130.0          161.0                        2.0   
99           101.0          197.0                        8.0   
100           72.0          165.0                        3.0   

     histogram_num

_[Your justification for 1.3 goes here]_

### Problem 1.4

In [98]:
### Your code for 1.4 goes here ###

### Problem 1.5

In [99]:
### Your code for 1.5 goes here ###

### Problem 1.6

In [100]:
### Your code for 1.6 goes here ###

# **2. Heart Failure Prediction Dataset**

## Problem 2.1

In [101]:
# load the dataset csv file into a pandas dataframe

### heart_df = pd.read_csv('heart_failure_dataset.csv') ###

### (a)

_[Your answer for 2.1(a) goes here]_

### (b)

In [102]:
### Your code for 2.1(b) goes here ###

### (c)



In [103]:
### Your code for 2.1(c) goes here ###

### (d)

In [104]:
### Your code for 2.1(d) goes here ###

### (e)

_[Your answer for 2.1(e) goes here]_

### (f)

_[Your answer for 2.1(f) goes here]_

### (g)

_[Your answer for 2.1(g) goes here]_

### (h)

_[Your answer for 2.1(h) goes here]_

### (i)

_[Your answer for 2.1(i) goes here]_

### (j)

In [105]:
### Your code for 2.1(j) goes here ###

_[Your justification for 2.1(j) goes here]_

### (k)

_[Your answer for 2.1(k) goes here]_

### (l)

In [106]:
### Your code for 2.1(l) goes here ###

### (m)

In [107]:
### Your code for 2.1(m) goes here ###

### (n)

In [108]:
### Your code for 2.1(n) goes here ###

## Problem 2.2

In [109]:
### Your code for 2.2 goes here ###

_[Your justification for 2.2 goes here]_

## Problem 2.3

In [110]:
### Your code for 2.3 goes here ###

## Problem 2.4

In [111]:
### Your code for 2.4 goes here ###

## Problem 2.5

In [112]:
### Your code for 2.5 goes here ###

### (a)

_[Your answer for 2.5(a) goes here]_

### (b)

_[Your answer for 2.5(b) goes here]_

### (c)

_[Your answer for 2.5(c) goes here]_

### (d)

_[Your answer for 2.5(d) goes here]_